### DISCUSSION TITLE
Using Method 1, can I add additional VIPs to pgbouncer?

### DISCUSSION THREAD

**Discussion by z7x** [autobase-tech/autobase#605] [General]
So, I managed to set up three Ubuntu VMs thanks to @vitabaks (thank you) with the following IPs:

Site 1
10.2.7.44
10.2.7.45

Site 2
192.168.16.46

The VIP is 10.2.7.48, but if Site 1 goes down (i.e. due to a power outage), I'd like to have a second VIP, say 192.168.16.48, to be waiting for connections.

Is this possible?

Can I simply edit vip-manager.yml and change the IP to 192.168.16.48 on Site 2? It seems to be switching over correctly, but do I use DNS round-robin to work with this?

**Comment by vitabaks** [comment]
Hello @z7x 

The **vip-manager** is designed to run the VIP on the server where the leader (Primary) Postgres role is, so it is expected that it will be one subnet.

In your case, there may be [PostgreSQL High-Availability with HAProxy Load Balancing](https://github.com/vitabaks/postgresql_cluster?tab=readme-ov-file#2-postgresql-high-availability-with-haproxy-load-balancing) scheme using HAProxy and **keepalived** (to provide an VIP address). In our configuration keepalived checks the status of the HAProxy service and in case of a failure delegates the VIP to another balancer server.  That is, VIP is no longer tied to the leader role.

I've already had a similar discussion on the Slack channel (available to project [sponsors](https://github.com/vitabaks/postgresql_cluster?tab=readme-ov-file#sponsor-this-project)), I'll duplicate it for you:


> Given this setup, I’m considering the implementation of two separate sets of HAProxy load balancers, one for each data center, with each set having its own VIP within the local IP range.

I think this is possible if you set `cluster_vip` variable for each region. But since now the variable is set at the vars/main level, that is, it is common, I can recommend using a separate inventory file for each data center, which will have its own set of nodes in the balancers group, and a given nested cluster_vip in the `[all:vars]` section, or you can define a section `[balancers:vars`] and set `cluster_vip` there.

**Reply by z7x** [reply]
Hi @vitabaks 

Thanks for your reply. Can I re-run the "deploy" command with Method 2 to change from pgboucner to HAproxy load balancer and then add additional VIP instances?

Are you open to doing some paid work for me, as I am not an expert :) (I didn't know how to message you privately, so please feel free to remove this line from here when you see it)

**Reply by vitabaks** [reply]
> Can I re-run the "deploy" command with Method 2 to change from pgboucner to HAproxy load balancer and then add additional VIP instances?

```
ansible-playbook balancers.yml -e "vip_manager_disable=true"
```

Note: Here we have added the vip_manager_disable variable to disable it

> Are you open to doing some paid work for me

Please see the [sponsoring](https://github.com/vitabaks/postgresql_cluster#sponsor-this-project) program to get personalized support.

**Comment by z7x** [comment]
> I think this is possible if you set `cluster_vip` variable for each region. But since now the variable is set at the vars/main level, that is, it is common, I can recommend using a separate inventory file for each data center, which will have its own set of nodes in the balancers group, and a given nested cluster_vip in the `[all:vars]` section, or you can define a section `[balancers:vars`] and set `cluster_vip` there.

Would you mind elaborating on this?

How do I deploy individual inventory to each site separately, ensuring the replication is running between all servers, whilst keeping only one leader server and two HAproxy/load balancers?

Many thanks

**Reply by vitabaks** [reply]
Use 3 separate inventory files for this task. First, deploy the database cluster, and then the load balancers separately for each site.

### Database cluster

Note: Leave the 'with_haproxy_load_balancing' variable with the value 'false', and comment out the 'cluster_vip' variable in vars/main.yml.

`cluster.hosts`
```
[etcd_cluster]
10.2.7.xx
192.168.16.xx
10.128.50.xx

# PostgreSQL nodes
[master]
10.2.7.44 hostname=pgnode01 postgresql_exists=false

[replica]
10.2.7.45 hostname=pgnode02 postgresql_exists=false
192.168.16.46 hostname=pgnode03 postgresql_exists=false

[postgres_cluster:children]
master
replica

[all:vars]
ansible_connection='ssh'
ansible_ssh_port='22'
ansible_user='root'
ansible_ssh_pass='secretpassword'
#ansible_ssh_private_key_file=
```

Note: **It is assumed that the ETCD cluster is located in 3 sites**, see examples [here](https://www.cybertec-postgresql.com/en/introduction-and-how-to-etcd-clusters-for-patroni/)

Deploy a database cluster:
`ansible-playbook deploy_pgcluster.yml -i cluster.hosts`

### Load balancers (site A)

`balancers.site_a.hosts`
```
[etcd_cluster]
10.2.7.xx
192.168.16.xx
10.128.50.xx

[postgres_cluster]
# listing of database servers not required

[balancers]
10.2.7.44
10.2.7.45

[balancers:vars]
cluster_vip='10.2.7.48'

[all:vars]
ansible_connection='ssh'
ansible_ssh_port='22'
ansible_user='root'
ansible_ssh_pass='secretpassword'
#ansible_ssh_private_key_file=
```

Note: Here we define the ETCD nodes of the cluster because the `confd` service (_responsible for managing the HAProxy configuration_) will automatically receive a list of Postgres servers from the DCS. But we leave an empty `postgres_cluster` group for the load balancer inventory file just to avoid an error when creating the HAProxy configuration file. P.S. I will fix this in the future so that we can omit the postgres_cluster group.

Deploy the load balancers for site A

`ansible-playbook balancers.yml -i balancers.site_a.hosts`

### Load balancers (site B)

`balancers.site_b.hosts`
```
[etcd_cluster]
10.2.7.xx
192.168.16.xx
10.128.50.xx

[balancers]
192.168.16.46

[balancers:vars]
cluster_vip='192.168.16.48'

[postgres_cluster]
# listing of database servers not required

[all:vars]
ansible_connection='ssh'
ansible_ssh_port='22'
ansible_user='root'
ansible_ssh_pass='secretpassword'
#ansible_ssh_private_key_file=
```

Deploy the load balancers for site B

`ansible-playbook balancers.yml -i balancers.site_b.hosts`

Note: In these examples, the load balancers are located on database servers, but they can also be dedicated servers. A minimum of 2 load balancer servers per site is recommended. therefore, you could think about dedicated VMs for load balancers, or add second VM with a database for the second site and install the load balancer on servers with databases.

**Reply by z7x** [reply]
Thanks, @vitabaks, I successfully deployed all three inventories with the following output for the first inventory file (no errors in any of the three).

I ended up creating 4 Ubuntu server VMs on Site A and Site B. The IPs are different this time.

![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/52f47c0a-8c3b-4727-93e9-fc1fdda723ba)

The HAProxy log is this:
```
`Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u01 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 3 active and 0 b>Mar 21 22:03:52 localhost haproxy[13000]: [WARNING]  (13000) : Server replicas_sync/u01 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 1>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u01 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 11ms. 3 active and 0 b>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 14ms. 2 active and 0 b>Mar 21 22:03:52 localhost haproxy[13000]: [WARNING]  (13000) : Server replicas_sync/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 1>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 14ms. 2 active and 0 b>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u03 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 2ms. 1 active and 0 ba>Mar 21 22:03:52 localhost haproxy[13000]: [WARNING]  (13000) : Server replicas_sync/u03 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 2>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u03 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 2ms. 1 active and 0 ba>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u04 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 1ms. 0 active and 0 ba>Mar 21 22:03:52 localhost haproxy[13000]: [WARNING]  (13000) : Server replicas_sync/u04 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 1>Mar 21 22:03:52 localhost haproxy[13000]: Server replicas_sync/u04 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 1ms. 0 active and 0 ba>Mar 21 22:03:52 localhost haproxy[13000]: proxy replicas_sync has no server available!`

I can connect to each PSQL server individually via the 5432 or 6432 ports; I can only make changes to the leader. At this point, I am unsure which IPs are the VIPs for each site.

Exit:

From what I can see in `haproxy.log`, some entries say the server X is UP and then down.

`Mar 21 23:07:46 localhost haproxy[1475]: proxy replicas_sync has no server available!                                                                                                   Mar 21 23:07:46 localhost haproxy[1475]: Server replicas_async/u01 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 12ms. 3 active and 0 b>Mar 21 23:07:46 localhost haproxy[1475]: [WARNING]  (1475) : Server replicas_async/u01 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 12>Mar 21 23:07:46 localhost haproxy[1475]: Server replicas_async/u01 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 12ms. 3 active and 0 b>Mar 21 23:07:46 localhost haproxy[1475]: Server replicas_async/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 13ms. 2 active and 0 b>Mar 21 23:07:46 localhost haproxy[1475]: [WARNING]  (1475) : Server replicas_async/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 13>Mar 21 23:07:46 localhost haproxy[1475]: Server replicas_async/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 13ms. 2 active and 0 b>Mar 21 23:07:49 localhost haproxy[1475]: Server replicas/u02 is UP, reason: Layer7 check passed, code: 200, check duration: 13ms. 3 active and 0 backup servers online. 0 sessions requ>Mar 21 23:07:49 localhost haproxy[1475]: [WARNING]  (1475) : Server replicas/u02 is UP, reason: Layer7 check passed, code: 200, check duration: 13ms. 3 active and 0 backup servers onl>Mar 21 23:07:49 localhost haproxy[1475]: Server replicas/u02 is UP, reason: Layer7 check passed, code: 200, check duration: 13ms. 3 active and 0 backup servers online. 0 sessions requ>Mar 21 23:07:50 localhost haproxy[1475]: Server replicas_async/u02 is UP, reason: Layer7 check passed, code: 200, check duration: 18ms. 3 active and 0 backup servers online. 0 session>Mar 21 23:07:50 localhost haproxy[1475]: [WARNING]  (1475) : Server replicas_async/u02 is UP, reason: Layer7 check passed, code: 200, check duration: 18ms. 3 active and 0 backup serve>
   `
```
Edit2:

It seems I can connect using port 5000 to all 4 servers and make changes on all of them (is this expected?), but the VIPs I set for both sites, 10.2.7.39 and 192.168.16.39, are currently dead.

**Reply by vitabaks** [reply]
In this case, given your "non-standard" configuration, do not pay attention to the output of "PostgreSQL Cluster connection info". Use port 5000 and the load balancer will direct traffic to the Primary server.

Details [here](https://github.com/vitabaks/postgresql_cluster?tab=readme-ov-file#2-postgresql-high-availability-with-haproxy-load-balancing) \
I recommend reading the README, there is a lot of useful information (including links)

**Reply by vitabaks** [reply]
> but the VIPs I set for both sites, 10.2.7.39 and 192.168.16.39, are currently dead.

Can I get the playbook execution log for the load balancer? 

check the list of IP addresses on balancer servers (`ip a`) and the **keepalived** service log.

**Reply by vitabaks** [reply]
I think I understood what was going on, the (empty) `cluster_vip` variable set at the `vars/main.yml` level redefined the variable we set at the inventory level. And therefore, we most just did not execute the keepalived role. \
So just comment out the cluster_vip variable in the main.yml and run the `ansible-playbook balancers.yml ...` command again.

**Comment by z7x** [comment]
> I think I understood what was going on, the (empty) `cluster_vip` variable set at the `vars/main.yml` level redefined the variable we set at the inventory level. And therefore, we most just did not execute the keepalived role. So just comment out the cluster_vip variable in the main.yml and run the `ansible-playbook balancers.yml ...` command again.

#### Thank you so much. This worked, and all _seems_ to be working as expected. 😎 (I think)

Can I check with you to see if I enabled SSL correctly?

I went to pgbouncer.ini on each server to manually specify the parameters as below.
```
client_tls_sslmode = prefer
client_tls_key_file = /etc/pgbouncer/server.cer
client_tls_cert_file = /etc/pgbouncer/server.key
```
I obtained this wildcard SSL from my IONOS account and I placed it in pgbouncer directory.
I tried using SSL generated by Certbot through Let's Encrypt, copied them to /etc/pgbouncer/, but pgbouncer would fail to load. I wanted to see if LetsEncrypt SSL would work before running `ansible-playbook` but pgbouncer doesn't seem like them for some reason.
I tried with: fullchain.pem and privkey.pem pair, chain.pem and/or cert.pem.
Certbot generates an SSL for a single domain and I have certbot running on each server.

I wonder if there are any parameters I need to specify, either in certbot or pgbouncer. I'm looking into this.
Not a big deal as the main thing is working but it would be great to have the SSL renewed automatically 😋

**Reply by vitabaks** [reply]
At the moment, TLS is not automatically configured, but it is in the plans - https://github.com/vitabaks/postgresql_cluster/issues/361

**Reply by vitabaks** [reply]
See pgbouncer variables in `vars/main.ym` (related PR https://github.com/vitabaks/postgresql_cluster/pull/585)

```yaml
pgbouncer_client_tls_sslmode: "disable"
pgbouncer_client_tls_key_file: ""
pgbouncer_client_tls_cert_file: ""
```

This is preferable to manual actions.

**Reply by vitabaks** [reply]
Manual changes are not desirable. For example, you specified `pgbouncer_client_tls_cert_file` instead of `client_tls_cert_file` in pgbouncer.ini file.

**Reply by z7x** [reply]
This was just a typo when I was replying to your message. 👍🏼

If I run `ansible-playbook` with a non-working SSL cert, will all servers (pgbouncer) potentially be down? I think this was the only reason why I applied it manually on one server first to see if it worked. Once it worked I just went and manually adjusted the .ini file in the other three servers, I'm using MobaXterm which allows "multi-execution" mode.

Otherwise, so far so good. I imported a table with 530k rows through a DNS hostname pointing to both VIPs. All worked really well!

**Comment by z7x** [comment]
Hi @vitabaks 

I noticed a problem with my setup. I took Site B down (both VMs) to test, and I could no longer connect via Site A VIP until at least one of the VMs at Site B was running and vice versa. I can't connect at all if one of the sites is completely down.

Any ideas what could have gone wrong? Could Apache2 interfere with any of these components? My colleague installed it a few days ago on all 4 servers to build custom PHP API access. Although, now thinking about it, I don't think I actually tested taking one site down when I thought it was all working.

These are the recent logs
for HAproxy:
```
Mar 29 15:03:03 localhost haproxy[5770]: proxy replicas_sync has no server available!
Mar 29 15:03:03 localhost haproxy[5770]: [NOTICE]   (5770) : haproxy version is 2.4.24-0ubuntu0.22.04.1
Mar 29 15:03:03 localhost haproxy[5770]: [NOTICE]   (5770) : path to executable is /usr/sbin/haproxy
Mar 29 15:03:03 localhost haproxy[5770]: [ALERT]    (5770) : proxy 'replicas_sync' has no server available!
Mar 29 15:03:03 localhost haproxy[5770]: proxy replicas_sync has no server available!
Mar 29 15:03:03 localhost haproxy[5770]: Server replicas_async/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 4ms. 3 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Mar 29 15:03:03 localhost haproxy[5770]: [WARNING]  (5770) : Server replicas_async/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 4ms. 3 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Mar 29 15:03:03 localhost haproxy[5770]: Server replicas_async/u02 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 4ms. 3 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Mar 29 15:03:03 localhost haproxy[5770]: Server replicas_async/u03 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 28ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Mar 29 15:03:03 localhost haproxy[5770]: [WARNING]  (5770) : Server replicas_async/u03 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 28ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Mar 29 15:03:03 localhost haproxy[5770]: Server replicas_async/u03 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 28ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
Mar 29 15:03:07 localhost haproxy[5770]: Server replicas/u03 is UP, reason: Layer7 check passed, code: 200, check duration: 21ms. 3 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.
Mar 29 15:03:07 localhost haproxy[5770]: [WARNING]  (5770) : Server replicas/u03 is UP, reason: Layer7 check passed, code: 200, check duration: 21ms. 3 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.
Mar 29 15:03:07 localhost haproxy[5770]: Server replicas/u03 is UP, reason: Layer7 check passed, code: 200, check duration: 21ms. 3 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.
Mar 29 15:03:07 localhost haproxy[5770]: Server replicas_async/u03 is UP, reason: Layer7 check passed, code: 200, check duration: 20ms. 3 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.
Mar 29 15:03:07 localhost haproxy[5770]: [WARNING]  (5770) : Server replicas_async/u03 is UP, reason: Layer7 check passed, code: 200, check duration: 20ms. 3 active and 0 backup servers online. 0 sessions requeued, 0 total in queue.

```

ETCD service status:
![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/a79f86fe-08b9-43a3-9fea-790040b11b92)
![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/47853fe1-90c0-4c05-a817-2726bb67ead1)
![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/76149151-f655-47d1-b18e-22d56df235e2)
![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/7403a38c-ce5d-44b3-af20-cdfacd376d4e)

Keepalived service status:
![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/8f2fd1f8-b809-4476-87e8-e7b09307147a)

**Reply by vitabaks** [reply]
Hi @z7x 

please attach the keepalived configuration and log (in text format, not screenshots).

**Reply by vitabaks** [reply]
> Note: It is assumed that the ETCD cluster is located in 3 sites, see examples [here](https://www.cybertec-postgresql.com/en/introduction-and-how-to-etcd-clusters-for-patroni/)

Have you read this? What is your DCS (etcd) cluster topology?

**Reply by z7x** [reply]
Hopefully, syslog is what you need. Please see attached.
[syslog.zip](https://github.com/vitabaks/postgresql_cluster/files/14808826/syslog.zip)

**Reply by z7x** [reply]
> > Note: It is assumed that the ETCD cluster is located in 3 sites, see examples [here](https://www.cybertec-postgresql.com/en/introduction-and-how-to-etcd-clusters-for-patroni/)
> 
> Have you read this? What is your DCS (etcd) cluster topology?

@vitabaks are you saying I should have an odd number of servers? For example, 3 or 5 and not 2 or 4 or 6?
Would you suggest that I remove or add one?

I'm happy to remove one from SIte B - if so, how do I remove one node from the cluster? 🤔

Or... should I have 3 nodes in Site A and 1 on SIte B?

I have no option to introduce Site C (i.e. Azure) and I'd prefer to avoid it.

**Reply by vitabaks** [reply]
Please see https://www.cybertec-postgresql.com/en/introduction-and-how-to-etcd-clusters-for-patroni/

**Reply by vitabaks** [reply]
About RAFT consensus algorithm http://thesecretlivesofdata.com/raft/

**Reply by vitabaks** [reply]
> I have no option to introduce Site C (i.e. Azure) and I'd prefer to avoid it.

Standby Cluster may be suitable for you. This is when two sites have two separate clusters etcd and patroni, where the second (standby) cluster is replicated from the main cluster. 

In this configuration, there is no automatic failover to the second site, it serves for disaster recovery (DR).

Details https://patroni.readthedocs.io/en/latest/replica_bootstrap.html#standby-cluster

**Reply by z7x** [reply]
Can I check the following in order to set up a standby cluster?

Site A inventory

```
[etcd_cluster]  # recommendation: 3, or 5-7 nodes
10.2.7.42
10.2.7.43

# PostgreSQL nodes
[master]
10.2.7.43 hostname=u04 postgresql_exists=false

[replica]
10.2.7.42 hostname=u03 postgresql_exists=false

[postgres_cluster:children]
master
replica

[balancers]
10.2.7.42
10.2.7.43

[balancers:vars]
cluster_vip='10.2.7.39'


[all:vars]
ansible_connection='ssh'
ansible_ssh_port='22'
ansible_user='root'
ansible_ssh_pass='password'
```

Site B inventory

```
[etcd_cluster]  # recommendation: 3, or 5-7 nodes
192.168.16.40
192.168.16.41

# PostgreSQL nodes
[master]
192.168.16.40 hostname=u01 postgresql_exists=false

[replica]
192.168.16.41 hostname=u02 postgresql_exists=false

[postgres_cluster:children]
master
replica

[balancers]
192.168.16.40
192.168.16.41

[balancers:vars]
cluster_vip='192.168.16.39'


[all:vars]
ansible_connection='ssh'
ansible_ssh_port='22'
ansible_user='root'
ansible_ssh_pass='password'
```

In terms of `vars/main.yml`, do I need `main_site_a.yml` and `main_site_b.yml` and specify unique name for each cluster so it should look as follows?:

Site A main.yml

```
patroni_cluster_name: "postgres-main-cluster"
patroni_standby_cluster:
  host: "192.168.16.40"  # an address of remote master
  port: "5432"  # a port of remote master
```

Site B main.yml

```
patroni_cluster_name: "postgres-standby-cluster"
patroni_standby_cluster:
  host: "10.2.7.43"  # an address of remote master
  port: "5432"  # a port of remote master
```

Did I miss anything or did I get it wrong?
Can I re-run this setup on the existing setup I've done last week? Apart from setting `postgresql_exists=true`, is there anything else I need to amend/comment out?

I can use REST API to promote the Standby Cluster to become the new Master Cluster when the original master goes down and then back to the original master when it's back online. (according to their documentation, it should work with our cloud workflows)

**Reply by vitabaks** [reply]
> In terms of vars/main.yml, do I need main_site_a.yml and main_site_b.yml and specify unique name for each cluster so it should look as follows?:

Yes, you can specify any cluster name at your discretion.

> Can I re-run this setup on the existing setup I've done last week? Apart from setting postgresql_exists=true

Yes it should work.

> I can use REST API to promote the Standby Cluster to become the new Master Cluster when the original master goes down and then back to the original master when it's back online. 

After the backup cluster is promoted (if the cluster is lost on site A) You need to create a new cluster on site A and make it a Standby cluster.

Before promote:

`Site A (Main cluster) --> Site B (Standby cluster)`

After promote:

`Site A (Standby cluster) <-- Site B (Main cluster)`

**Reply by vitabaks** [reply]
In your inventory, I see that only two clusters are listed in the "etcd_cluster" group, at least 3 servers are required for the DCS cluster. You can deploy an etcd cluster on separate 3 small servers and use it for multiple Postgres clusters where there may already be 2 servers in a database cluster.

It is important to understand that the DCS cluster and the Postgres cluster are different clusters. The requirement for a minimum of 3 servers applies to the DCS cluster.

Please see:
- https://patroni.readthedocs.io/en/latest/ha_multi_dc.html
-  http://thesecretlivesofdata.com/raft/

**Reply by vitabaks** [reply]
you need “manually promote” the healthy cluster by removing standby_cluster section from the [dynamic configuration](https://patroni.readthedocs.io/en/latest/dynamic_configuration.html#dynamic-configuration).

Promote standby cluster via patronictl
```
patronictl edit-config -s standby_cluster=null --force
```

**Comment by z7x** [comment]
> Hi @z7x
> 
> please attach the keepalived configuration and log (in text format, not screenshots).



> > Note: It is assumed that the ETCD cluster is located in 3 sites, see examples [here](https://www.cybertec-postgresql.com/en/introduction-and-how-to-etcd-clusters-for-patroni/)
> 
> Have you read this? What is your DCS (etcd) cluster topology?

This is my setup atm
![image](https://github.com/vitabaks/postgresql_cluster/assets/617043/87734823-129a-42fd-b195-1acdc5940aab)

This was my idea of my configuration. The subdomain can be controlled/updated by our custom API cloud workflow and A records updated/added/removed etc.

**Reply by vitabaks** [reply]
`etcdctl member list`

**Reply by z7x** [reply]
> `etcdctl member list`
10a14a7982ffa573, started, u01, http://192.168.16.40:2380, http://192.168.16.40:2379, false
1cbfb298bc2bfe33, started, u03, http://10.2.7.42:2380, http://10.2.7.42:2379, false
5bbf39ff9128f403, started, u02, http://192.168.16.41:2380, http://192.168.16.41:2379, false
a877dfb56c236613, started, u04, http://10.2.7.43:2380, http://10.2.7.43:2379, false
